src/Simulation/Common/QubitManager.cs

```diff
@@ -120,23 +121,37 @@ public virtual void OnOperationEnd(ICallable _, IApplyData values)
             }
         }
 
-        protected IEnumerable<Qubit> QubitsInUse(StackFrame frame)
+        protected HashSet<Qubit> QubitsInUse(StackFrame frame)
         {
             if (DisableBorrowing || frame == null)
             {
-                return Qubit.NO_QUBITS;
+                return EMPTY_SET;
             }
 
-            return frame.Locals.Keys.Concat(frame.QubitsInArgument ?? Array.Empty<Qubit>())
-                .Where(q => q != null && !this.IsDisabled(q));
+            HashSet<Qubit> inUse = new HashSet<Qubit>();
```

Comment:
Would it make sense to pre-allocate a little? Maybe size of locals?

Reply:
yep, good idea